README.md
![Labcodeset transform build](https://github.com/aehrc/labcodeset-fhir-transform/workflows/Labcodeset%20transform%20build/badge.svg)

# Labcodeset to FHIR transformation

A stand-alone command-line tool that transforms Nederlandse Labcodeset XML to a collection of FHIR resources, specifically

| Type | URI | Description |
| ---- | --- | ----------- |
| CodeSystem supplement | http://labterminologie.nl/cs/labconcepts | CodeSystem supplement containing Dutch translations and properties as defined in the Labcodeset XML file |
| ValueSet | http://labterminologie.nl/vs/labconcepts | ValueSet containing the unique set of LOINC codes referenced by the Labcodeset XML file - note this contains LOINC codes in any state, not just active codes |
| CodeSystem fragment | http://unitsofmeasure.org | Fragment of the UCUM CodeSystem containing the union of UCUM expressions referenced by Labcodeset and the ["common" FHIR UCUM ValueSet](https://www.hl7.org/fhir/valueset-ucum-common.html). Language for this CodeSystem is nl_NL, and Dutch display terms are used where available with the English terms set as designations |
| ValueSet | http://labterminologie.nl/vs/labconcepts-ucum | ValueSet containing the unique set of UCUM expressions referenced by Labcodeset |
| ConceptMap | http://labterminologie.nl/cm/labconcepts-ucum | ConceptMap containing mappings for each LOINC code in Labcodeset where a unit is expressed |
| ValueSet | http://labterminologie.nl/vs/labconcepts-materials | ValueSet containing the unique set of SNOMED CT "material" codes referenced by Labcodeset |
| ConceptMap | http://labterminologie.nl/cm/labconcepts-materials | ConceptMap containing mappings from LOINC codes in Labcodeset that reference one or more SNOMED CT "material" codes |
| ValueSet | http://labterminologie.nl/vs/labconcepts-ordinal-xxx | A set of ValueSets, one for each "ordinal" list included in the Labcodeset XML file where xxx in the URI is replaced with the OID of the ordinal set |
| ConceptMap | http://labterminologie.nl/cm/labconcepts-outcomes | ConceptMap containing mappings from LOINC codes in Labcodeset to either a SNOMED CT Reference Set identifier or a ValueSet OID depending upon which is listed in Labcodeset for that LOINC code.
 

## Build
>mvn clean package

This will build a file called ./target/labcodeset-transforms-jar-with-dependencies.jar

If you just want to run it, you can simply download the labcodeset-transforms-jar-with-dependencies.jar from the [prebuilt release packages](https://github.com/orgs/aehrc/packages?repo_name=labcodeset-fhir-transform) and skip to the next step.

## Run

The program requires a few command line parameters to execute, they are desribed by the usage message if you don't supply the correct parameters.

        usage: java -jar labcodeset-transforms-jar-with-dependencies.jar [-clientId <client_id>] [-clientSecret <client_secret>]
               -fhirEndpoint <fhir endpoint> -labcodesetFile <file_path> -loincVersion <loinc_version> [-outputDir <directory>]
               [-tokenEndpoint <token _endpoint>]
         -clientId <client_id>              Client id part of client credentials to use with the token endpoint to get a bearer
                                            token for use with the specified FHIR endpoin - optional if authorisation is
                                            required for the FHIR endpoint
         -clientSecret <client_secret>      Client secret part of client credentials to use with the token endpoint to get a
                                            bearer token for use with the specified FHIR endpoin  - optional if authorisation is
                                            required for the FHIR endpoint
         -fhirEndpoint <fhir endpoint>      FHIR terminology endpoint containing the appropriate LOINC and SNOMED CT-NL versions
                                            to reference during the transformation
         -labcodesetFile <file_path>        File path to the Labcodeset XML file to transform
         -loincVersion <loinc_version>      Version of LOINC this Labcodeset XML file has been built with
         -outputDir <directory>             Location to output the resulting FHIR resources and Bundle - defaults to the
                                            directory the program was executed from if not specified
         -tokenEndpoint <token _endpoint>   Token endpoint URL to get a bearer token with for the specified endpoint - optional
                                            if authorisation is required for the FHIR endpoint

Following is an example execution

        java -jar labcodeset-transforms-jar-with-dependencies.jar \
            -fhirEndpoint https://r4.ontoserver.csiro.au/fhir \
            -labcodesetFile /some/path/labcodeset.xml \
            -loincVersion 2.69

## Output
The transform will produce one JSON file for each of the generated resource as well as a Bundle resource JSON file containing all of the resources as a convenience.

The file names of the generated JSON files will contain the version number inside each of the resources, and the Bundle resource file name will carry the Labcodeset version it was generated from.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.1] - 2021-02-23
### Fixed
- Copyright statement for generated resources
- Logging dependency warning

## [1.0.0] - 2021-02-05
 - Initial version


src/main/java/au/csiro/fhir/transforms/utility/Constants.java
/*******************************************************************************
 * Copyright © 2021, Commonwealth Scientific and Industrial Research
 * Organisation (CSIRO) ABN 41 687 119 230. Licensed under the CSIRO Open Source
 * Software Licence Agreement.
 ******************************************************************************/
package au.csiro.fhir.transforms.utility;

/**
 * General constants used across the transformations
 */
public class Constants {
  /**
   * Official FHIR CodeSystem URI for LOINC
   */
  public static final String LOINC_CS_URI = "http://loinc.org";

  /**
   * Official FHIR CodeSystem URI of SNOMED CT
   */
  public static final String SCT_CS_URI = "http://snomed.info/sct";

  /**
   * URI of the Netherlands edition of SNOMED CT
   */
  public static final String NL_SCT_EDITION = "http://snomed.info/sct/11000146104";

  /**
   * OID CodeSystem URI for FHIR
   */
  public static final String OID_CS_URI = "urn:ietf:rfc:3986";

  /**
   * Official FHIR specification UCUM CodeSystem URI
   */
  public static final String UCUM_CS_URI = "http://unitsofmeasure.org";

  /**
   * Language code for Dutch used on all Labcodeset resources
   */
  public static final String DUTCH_LANGUAGE_CODE = "nl-NL";

  /**
   * UCUM version being used - fairly static
   */
  public static final String UCUM_VERSION = "2.1";

  /**
   * Publisher set on all Labcodeset generated resources
   */
  public static final String LABCODESET_RESOURCE_PUBLISHER = "Nictiz";

  /**
   * Copyright statement set on all Labcodeset generated resources
   */
  public static final String LABCODESET_RESOURCE_COPYRIGHT =
      "The Labcodeset is developed and maintained by NVKC (Nederlandse Vereniging voor Klinische Chemie en Laboratoriumgeneeskunde) and NVMM (Nederlandse Vereniging voor Medische Microbiologie), with functional maintenance by Nictiz (Nationaal ICT Instituut in de Zorg). If you have any questions or change requests, please email labcodeset@nictiz.nl.\n"
          + "The Labcodeset is based on LOINC, UCUM and SNOMED. Its use requires a license for the Dutch SNOMED Edition, which can be obtained at https://mlds.ihtsdotools.org/.\n"
          + "This material contains content from LOINC (http://loinc.org). LOINC is copyright © 1995-2020, Regenstrief Institute, Inc. and the Logical Observation Identifiers Names and Codes (LOINC) Committee and is available at no cost under the license at http://loinc.org/license. LOINC® is a registered United States trademark of Regenstrief Institute, Inc.\n"
          + "In general, Regenstrief’s terms-of-use for LOINC and UCUM have had the same core principles since its first release back in 1995. Certainly, the license has gotten longer and more explicit about certain things, but the main premises remain:\n"
          + "1. LOINC and UCUM are free for use in both commercial and non-commercial applications,\n"
          + "2. LOINC and UCUM are protected against change by end-users or creation of another vocabulary so that they can fulfil their goal of being an international standard, and\n"
          + "3. When LOINC or UCUM are used in databases, software applications, etc. users need to acknowledge the license and copyright so that end users of that product know that LOINC and UCUM are protected and licensed in a specific way.\n"
          + "The license agreements for LOINC and UCUM are available respectively at https://loinc.org/license/ and https://unitsofmeasure.org/trac/wiki/TermsOfUse";

  /**
   * Prefix used for URIs for Labcodeset generated resources
   */
  public static final String LABCODESET_URI_PREFIX = "http://labterminologie.nl";
}


src/main/java/au/csiro/fhir/transforms/utility/TerminologyClient.java
/*******************************************************************************
 * Copyright © 2021, Commonwealth Scientific and Industrial Research
 * Organisation (CSIRO) ABN 41 687 119 230. Licensed under the CSIRO Open Source
 * Software Licence Agreement.
 ******************************************************************************/
package au.csiro.fhir.transforms.utility;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.apache.http.HttpEntity;
import org.apache.http.NameValuePair;
import org.apache.http.client.ClientProtocolException;
import org.apache.http.client.entity.UrlEncodedFormEntity;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.message.BasicNameValuePair;
import org.apache.http.util.EntityUtils;
import org.hl7.fhir.r4.model.CodeSystem;
import org.hl7.fhir.r4.model.Parameters;
import org.hl7.fhir.r4.model.StringType;
import org.hl7.fhir.r4.model.UriType;
import org.hl7.fhir.r4.model.ValueSet;
import com.google.gson.JsonParser;
import ca.uhn.fhir.context.FhirContext;
import ca.uhn.fhir.rest.client.api.IGenericClient;
import ca.uhn.fhir.rest.server.exceptions.ResourceNotFoundException;

/**
 * Utility class to get information about code systems for the transformation
 */
public class TerminologyClient {

  /**
   * FHIR specification "common" UCUM codes ValueSet URL
   */
  private static final String UCUM_COMMON_CODES_VALUESET_URL = "https://www.hl7.org/fhir/valueset-ucum-common.json";

  private Map<String, String> snomedCache = new HashMap<>();
  private Map<String, Parameters> loincCache = new HashMap<>();
  private IGenericClient fhirClient;

  /**
   * @param fhirClient connection to a FHIR server to refer to
   */
  public TerminologyClient(IGenericClient fhirClient) {
    this.fhirClient = fhirClient;
  }

  /**
   * Looks up a LOINC code and returns a {@link Parameters} object from the FHIR server in response to
   * a $lookup for all properties of the code
   * 
   * @param code LOINC code to look up
   * @param loincVersion LOINC version to use
   * @return {@link Parameters} object returned by the FHIR server requesting all properties be
   *         returned
   */
  public Parameters getLoincConcept(String code, String loincVersion) {
    if (!loincCache.containsKey(code)) {
      Parameters inParams = new Parameters();
      inParams.addParameter().setName("code").setValue(new StringType(code));
      inParams.addParameter().setName("system").setValue(new UriType(Constants.LOINC_CS_URI));
      inParams.addParameter().setName("version").setValue(new StringType(loincVersion));
      inParams.addParameter().setName("property").setValue(new StringType("*"));
      loincCache.put(code,
          fhirClient.operation().onType(CodeSystem.class).named("$lookup").withParameters(inParams).useHttpGet().execute());
    } else {
      System.out.println("loinc cache hit");
    }

    return loincCache.get(code);
  }

  /**
   * Gets the preferred display term for a SNOMED CT code from the Netherlands edition
   * 
   * @param code SNOMED CT code to look up
   * @param defaultIfNotFound default text to return if the code is not found
   * @return the display term for the SNOMED CT code specified in the Netherlands edition of SNOMED
   *         CT, or the value passed as defaultIfNotFound if no SNOMED CT concept can be found for
   *         that code
   */
  public String getSnomedDisplay(String code, String defaultIfNotFound) {
    if (!snomedCache.containsKey(code)) {
      Parameters inParams = new Parameters();
      inParams.addParameter().setName("code").setValue(new StringType(code));
      inParams.addParameter().setName("system").setValue(new UriType(Constants.SCT_CS_URI));
      inParams.addParameter().setName("version").setValue(new StringType(Constants.NL_SCT_EDITION));
      inParams.addParameter().setName("property").setValue(new StringType("display"));
      String display;
      try {
        display = fhirClient.operation().onType(CodeSystem.class).named("$lookup").withParameters(inParams).useHttpGet().execute()
            .getParameters("display").get(0).primitiveValue();
      } catch (ResourceNotFoundException e) {
        System.err.println(
            "WARNING: SNOMED CT concept " + code + " not found, using the display term from the XML file '" + defaultIfNotFound + "'");
        display = defaultIfNotFound;
      }
      snomedCache.put(code, display);
    }

    return snomedCache.get(code);

  }

  /**
   * @return the FHIR specification common UCUM codes ValueSet
   * @throws IOException
   * @throws ClientProtocolException
   */
  public static ValueSet getCommonUcumCodes() throws ClientProtocolException, IOException {

    HttpGet request = new HttpGet(UCUM_COMMON_CODES_VALUESET_URL);
    try (CloseableHttpClient httpClient = HttpClients.createDefault(); CloseableHttpResponse response = httpClient.execute(request)) {

      // Get HttpResponse Status
      System.out.println(response.getProtocolVersion()); // HTTP/1.1

      if (response.getStatusLine().getStatusCode() != 200) {
        throw new RuntimeException("Unable to get common UCUM codes from the FHIR specification, response code was "
            + response.getStatusLine().getStatusCode() + " response was " + EntityUtils.toString(response.getEntity()));
      }

      HttpEntity entity = response.getEntity();
      if (entity != null) {
        return FhirContext.forR4().newJsonParser().parseResource(ValueSet.class, EntityUtils.toString(entity));
      } else {
        throw new RuntimeException("Unexpected emtpy response for " + UCUM_COMMON_CODES_VALUESET_URL);
      }
    }
  }

  public static String getToken(String tokenEndpoint, String clientId, String clientSecret) throws IOException {

    HttpPost post = new HttpPost(tokenEndpoint);

    List<NameValuePair> urlParameters = new ArrayList<>();
    urlParameters.add(new BasicNameValuePair("grant_type", "client_credentials"));
    urlParameters.add(new BasicNameValuePair("client_id", clientId));
    urlParameters.add(new BasicNameValuePair("client_secret", clientSecret));

    post.setEntity(new UrlEncodedFormEntity(urlParameters));

    try (CloseableHttpClient httpClient = HttpClients.createDefault(); CloseableHttpResponse response = httpClient.execute(post)) {
      return JsonParser.parseString(EntityUtils.toString(response.getEntity())).getAsJsonObject().get("access_token").getAsString();
    }
  }

}


src/main/java/au/csiro/fhir/transforms/generators/UcumResourceGenerator.java
/*******************************************************************************
 * Copyright © 2021, Commonwealth Scientific and Industrial Research
 * Organisation (CSIRO) ABN 41 687 119 230. Licensed under the CSIRO Open Source
 * Software Licence Agreement.
 ******************************************************************************/
package au.csiro.fhir.transforms.generators;

import java.io.IOException;
import java.util.HashSet;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import org.hl7.fhir.r4.model.CodeSystem;
import org.hl7.fhir.r4.model.CodeSystem.CodeSystemContentMode;
import org.hl7.fhir.r4.model.CodeSystem.ConceptDefinitionComponent;
import org.hl7.fhir.r4.model.ConceptMap;
import org.hl7.fhir.r4.model.ConceptMap.ConceptMapGroupComponent;
import org.hl7.fhir.r4.model.ConceptMap.SourceElementComponent;
import org.hl7.fhir.r4.model.ConceptMap.TargetElementComponent;
import org.hl7.fhir.r4.model.Enumerations.ConceptMapEquivalence;
import org.hl7.fhir.r4.model.Enumerations.PublicationStatus;
import org.hl7.fhir.r4.model.ValueSet;
import org.hl7.fhir.r4.model.ValueSet.ConceptReferenceComponent;
import org.hl7.fhir.r4.model.ValueSet.ConceptSetComponent;
import org.hl7.fhir.r4.model.ValueSet.ValueSetComposeComponent;
import au.csiro.fhir.transform.xml.nl.labcodeset.LabConcept;
import au.csiro.fhir.transform.xml.nl.labcodeset.LabConcept.Units;
import au.csiro.fhir.transform.xml.nl.labcodeset.MaterialDefinition;
import au.csiro.fhir.transform.xml.nl.labcodeset.Publication;
import au.csiro.fhir.transform.xml.nl.labcodeset.UnitDefinition;
import au.csiro.fhir.transforms.utility.Constants;
import au.csiro.fhir.transforms.utility.TerminologyClient;

public class UcumResourceGenerator {

  private static final String LABCODESET_UCUM_CM_DESCRIPTION = "Map from Nederlandse Labcodeset LOINC to UCUM units";
  private static final String LABCODESET_UCUM_CM_URI = Constants.LABCODESET_URI_PREFIX + "/cm/labconcepts-ucum";
  private static final String LABCODESET_UCUM_CM_TITLE = "Nederlandse Labcodeset UCUM ConceptMap";

  private static final String LABCODESET_UCUM_VS_TITLE = "Nederlandse UCUM";
  private static final String LABCODESET_UCUM_VS_URI = Constants.LABCODESET_URI_PREFIX + "/vs/labconcepts-ucum";
  private static final String LABCODESET_UCUM_VS_DESCRIPTION = "Nederlandse Labcodeset UCUM codes";

  private static final String UCUM_FRAGMENT_TITLE = "Unified Code for Units of Measure (UCUM)";
  private static final String UCUM_FRAGMENT_DESCRIPTION =
      "Fragment of the Unified Code for Units of Measure (UCUM) representing all commonly used UCUM codes specified in the "
          + "FHIR specification at https://www.hl7.org/fhir/valueset-ucum-common.html with additional Nederlandse Labcodeset expressions and Dutch translations";

  private String labcodesetVersion;
  private String loincVersion;
  private Map<String, UnitDefinition> unitMap;

  /**
   * @param labcodesetVersion version of the Labcodeset being transformed
   * @param loincVersion LOINC version the Labcodeset file should be used with
   * @param terminologyClient {@link TerminologyClient} that can be used to lookup details of LOINC
   *        codes
   * @param unitMap Map of the unit references and their details in the Labcodeset file
   * @param materialsMap Map of the materials references and their details in the Labcodeset file
   */
  public UcumResourceGenerator(String labcodesetVersion, String loincVersion, Map<String, UnitDefinition> unitMap,
      Map<String, MaterialDefinition> materialsMap) {
    this.labcodesetVersion = labcodesetVersion;
    this.loincVersion = loincVersion;
    this.unitMap = unitMap;
  }

  /**
   * @param pub Parsed Labcodeset XML object
   * @return {@link CodeSystem} resource containing the unique union of the Labcodeset UCUM
   *         expressions and expressions in the FHIR specification common UCUM codes with Dutch
   *         display names where available and English designations.
   */
  public CodeSystem createUcumCodeSystem(Publication pub) {
    CodeSystem codeSystem = new CodeSystem();
    String csIdentifier = "Ucum-" + Constants.UCUM_VERSION;
    codeSystem.setId(csIdentifier);
    codeSystem.setUrl(Constants.UCUM_CS_URI).setValueSet(Constants.UCUM_CS_URI + "/vs").setDescription(UCUM_FRAGMENT_DESCRIPTION)
        .setVersion(Constants.UCUM_VERSION).setTitle(UCUM_FRAGMENT_TITLE).setName(csIdentifier).setStatus(PublicationStatus.ACTIVE)
        .setExperimental(false).setContent(CodeSystemContentMode.FRAGMENT).setCopyright(Constants.LABCODESET_RESOURCE_COPYRIGHT)
        .setLanguage(Constants.DUTCH_LANGUAGE_CODE);

    Set<String> nlUcumCodeset = new HashSet<>();

    for (Entry<String, UnitDefinition> entry : unitMap.entrySet()) {
      UnitDefinition code = entry.getValue();
      if (!nlUcumCodeset.contains(code.getRm())) {
        ConceptDefinitionComponent ucumCode = codeSystem.addConcept();

        if (code.getRm() == null) {
          System.err.println("UCUM unit reference " + entry.getKey() + " has no defined UCUM expression - cannot continue");
          System.exit(1);
        }

        if (code.getNlname() == null) {
          System.err.println("UCUM unit reference " + entry.getKey() + " has no defined Dutch name - cannot continue");
          System.exit(1);
        }

        ucumCode.setCode(code.getRm());
        ucumCode.setDisplay(code.getNlname());

        if (code.getName() != null) {
          ucumCode.addDesignation().setLanguage("en").setValue(code.getName());
        } else {
          System.err.println("UCUM expression " + code.getRm() + " has no defined English name in Labcodeset");
        }

        nlUcumCodeset.add(code.getRm());
      }
    }

    ValueSet commonUcum = null;
    try {
      commonUcum = TerminologyClient.getCommonUcumCodes();
    } catch (IOException e) {
      System.err.println("Unable to get common UCUM codes from the FHIR spec due to " + e.getLocalizedMessage());
      System.exit(1);
    }

    for (ConceptSetComponent commonInclude : commonUcum.getCompose().getInclude()) {
      for (ConceptReferenceComponent concept : commonInclude.getConcept()) {
        if (!nlUcumCodeset.contains(concept.getCode())) {
          ConceptDefinitionComponent ucumCode = codeSystem.addConcept();
          ucumCode.setCode(concept.getCode());
          ucumCode.setDisplay(concept.getDisplay());

          nlUcumCodeset.add(concept.getCode());
        }
      }
    }

    return codeSystem;
  }

  /**
   * @param pub Parsed Labcodeset XML object
   * @return {@link ValueSet} containing the unique set of UCUM expressions referenced by Labcodeset
   */
  public ValueSet createUcumValueSet(Publication pub) {
    ValueSet valueSet = new ValueSet();
    String vsIdentifier = "Labconcepts-ucum-" + labcodesetVersion;
    valueSet.setId(vsIdentifier);
    valueSet.setName(vsIdentifier).setVersion(labcodesetVersion).setTitle(LABCODESET_UCUM_VS_TITLE).setStatus(PublicationStatus.ACTIVE)
        .setUrl(LABCODESET_UCUM_VS_URI).setDescription(LABCODESET_UCUM_VS_DESCRIPTION).setExperimental(false)
        .setPublisher(Constants.LABCODESET_RESOURCE_PUBLISHER).setCopyright(Constants.LABCODESET_RESOURCE_COPYRIGHT)
        .setLanguage(Constants.DUTCH_LANGUAGE_CODE);

    ValueSetComposeComponent component = new ValueSetComposeComponent();
    valueSet.setCompose(component);

    ConceptSetComponent include = component.addInclude();
    include.setSystem(Constants.UCUM_CS_URI);

    Set<String> nlUcumCodeset = new HashSet<>();

    for (UnitDefinition entry : unitMap.values()) {
      if (!nlUcumCodeset.contains(entry.getRm())) {
        ConceptReferenceComponent reference = include.addConcept();
        reference.setCode(entry.getRm());
        reference.setDisplay(entry.getNlname());

        nlUcumCodeset.add(entry.getRm());
      }
    }
    return valueSet;
  }

  /**
   * @param pub Parsed Labcodeset XML object
   * @return {@link ConceptMap} containing maps from LOINC codes to UCUM units as expressed in the
   *         references in Labcodeset
   */
  public ConceptMap createUcumMap(Publication pub) {
    ConceptMap map = new ConceptMap();

    String identifier = "Labconcepts-ucum-" + labcodesetVersion;
    map.setId(identifier);
    map.setTitle(LABCODESET_UCUM_CM_TITLE).setVersion(labcodesetVersion).setName(identifier).setStatus(PublicationStatus.ACTIVE)
        .setUrl(LABCODESET_UCUM_CM_URI).setDescription(LABCODESET_UCUM_CM_DESCRIPTION).setExperimental(false)
        .setPublisher(Constants.LABCODESET_RESOURCE_PUBLISHER).setCopyright(Constants.LABCODESET_RESOURCE_COPYRIGHT)
        .setLanguage(Constants.DUTCH_LANGUAGE_CODE);

    ConceptMapGroupComponent group = new ConceptMapGroupComponent();
    group.setSource(Constants.LOINC_CS_URI);
    group.setSourceVersion(loincVersion);
    group.setTarget(Constants.UCUM_CS_URI);
    group.setTargetVersion(Constants.UCUM_VERSION);

    for (LabConcept labConcept : pub.getLabConcepts().getLabConcept()) {
      Units units = labConcept.getUnits();
      if (units != null) {
        UnitDefinition unit = unitMap.get(units.getUnit().getRef());

        if (unit != null) {
          SourceElementComponent element = new SourceElementComponent();
          element.setCode(labConcept.getLoincConcept().getLoincNum());
          if (labConcept.getLoincConcept().getTranslation() != null
              && labConcept.getLoincConcept().getTranslation().getLongName() != null) {
            element.setDisplay(labConcept.getLoincConcept().getTranslation().getLongName().getValue());
          } else {
            element.setDisplay(labConcept.getLoincConcept().getLongName().getValue());
          }
          TargetElementComponent targetElement = new TargetElementComponent();
          element.addTarget(targetElement);
          targetElement.setEquivalence(ConceptMapEquivalence.RELATEDTO);

          targetElement.setCode(unit.getRm());
          targetElement.setDisplay(unit.getNlname());
          group.addElement(element);
        } else {
          System.err.println(
              "Unable to find unit for reference " + units.getUnit().getRef() + " on " + labConcept.getLoincConcept().getLoincNum());
        }
      }
    }

    map.addGroup(group);

    return map;

  }
}


src/main/java/au/csiro/fhir/transforms/LabcodesetFhirTransform.java
/*******************************************************************************
 * Copyright © 2021, Commonwealth Scientific and Industrial Research
 * Organisation (CSIRO) ABN 41 687 119 230. Licensed under the CSIRO Open Source
 * Software Licence Agreement.
 ******************************************************************************/
package au.csiro.fhir.transforms;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.CommandLineParser;
import org.apache.commons.cli.DefaultParser;
import org.apache.commons.cli.HelpFormatter;
import org.apache.commons.cli.Option;
import org.apache.commons.cli.Options;
import org.apache.commons.cli.ParseException;
import au.csiro.fhir.transforms.generators.LabcodesetResourceGenerator;

/**
 * Command line interface class for the Labcodeset transformation to FHIR
 */
public class LabcodesetFhirTransform {

  private static final String CLIENT_SECRET_PARAM = "clientSecret";
  private static final String CLIENT_ID_PARAM = "clientId";
  private static final String TOKEN_ENDPOINT_PARAM = "tokenEndpoint";
  private static final String OUTPUT_DIR_PARAM = "outputDir";
  private static final String FHIR_ENDPOINT_PARAM = "fhirEndpoint";
  private static final String LOINC_VERSION_PARAM = "loincVersion";
  private static final String LABCODESET_FILE_PARAM = "labcodesetFile";

  public static void main(String[] args) {
    // create the command line parser
    CommandLineParser parser = new DefaultParser();

    // create the Options
    Options options = new Options();
    options.addOption(Option.builder(LABCODESET_FILE_PARAM).required(true).argName("file_path").hasArg(true)
        .desc("File path to the Labcodeset XML file to transform").build());
    options.addOption(Option.builder(LOINC_VERSION_PARAM).required(true).argName("loinc_version").hasArg(true)
        .desc("Version of LOINC this Labcodeset XML file has been built with").build());
    options.addOption(Option.builder(FHIR_ENDPOINT_PARAM).required(true).argName("fhir endpoint").hasArgs().type(File.class)
        .desc("FHIR terminology endpoint containing the appropriate LOINC and SNOMED CT-NL versions to reference during the transformation")
        .build());
    options.addOption(Option.builder(TOKEN_ENDPOINT_PARAM).required(false).argName("token _endpoint").hasArg(true).desc(
        "Token endpoint URL to get a bearer token with for the specified endpoint - optional if authorisation is required for the FHIR endpoint")
        .build());
    options.addOption(Option.builder(CLIENT_ID_PARAM).required(false).argName("client_id").hasArg(true).desc(
        "Client id part of client credentials to use with the token endpoint to get a bearer token for use with the specified FHIR endpoin - optional if authorisation is required for the FHIR endpoint")
        .build());
    options.addOption(Option.builder(CLIENT_SECRET_PARAM).required(false).argName("client_secret").hasArg(true).desc(
        "Client secret part of client credentials to use with the token endpoint to get a bearer token for use with the specified FHIR endpoin  - optional if authorisation is required for the FHIR endpoint")
        .build());
    options.addOption(Option.builder(OUTPUT_DIR_PARAM).required(false).argName("directory").hasArg(true).desc(
        "Location to output the resulting FHIR resources and Bundle - defaults to the directory the program was executed from if not specified")
        .build());

    boolean initFailed = false;
    File labcodesetFile = null;
    String loincVersion = null;
    String fhirEndpoint = null;
    File outputDir = null;
    String tokenEndpoint = null;
    String clientId = null;
    String clientSecret = null;
    try {
      CommandLine line = parser.parse(options, args);

      labcodesetFile = new File(line.getOptionValue(LABCODESET_FILE_PARAM));

      loincVersion = line.getOptionValue(LOINC_VERSION_PARAM);
      if (!loincVersion.matches("\\d{1}\\.\\d{2,}")) {
        System.err.println("LOINC version specified " + loincVersion + " is not valid, expected x.yy");
        initFailed = true;
      }

      fhirEndpoint = line.getOptionValue(FHIR_ENDPOINT_PARAM);

      if (line.hasOption(OUTPUT_DIR_PARAM)) {
        outputDir = new File(line.getOptionValue(OUTPUT_DIR_PARAM));
        if (outputDir.exists() && (!outputDir.isDirectory() || !outputDir.canWrite())) {
          System.err.println("Output directory " + outputDir + " exists but is not a directory or not writable");
          initFailed = true;
        } else if (!outputDir.exists()) {
          try {
            Files.createDirectories(outputDir.toPath());
          } catch (IOException e) {
            System.err.println("Output directory " + outputDir + " does not exist but cannot be created " + e.getLocalizedMessage());
            initFailed = true;
          }
        }
      } else {
        outputDir = new File(System.getProperty("user.dir"));
      }

      if (options.hasOption(CLIENT_ID_PARAM) || options.hasOption(CLIENT_SECRET_PARAM) || options.hasOption(TOKEN_ENDPOINT_PARAM)) {
        if (options.hasOption(CLIENT_ID_PARAM) && options.hasOption(CLIENT_SECRET_PARAM) && options.hasOption(TOKEN_ENDPOINT_PARAM)) {
          tokenEndpoint = line.getOptionValue(TOKEN_ENDPOINT_PARAM);
          clientId = line.getOptionValue(CLIENT_ID_PARAM);
          clientSecret = line.getOptionValue(CLIENT_SECRET_PARAM);
        } else {
          System.err.println("Parameters " + TOKEN_ENDPOINT_PARAM + ", " + CLIENT_ID_PARAM + ", and " + CLIENT_SECRET_PARAM
              + " must either all be supplied together or none can be set");
          initFailed = true;
        }
      }

    } catch (ParseException e1) {
      System.err.println("Failed to parse arguments " + e1.getLocalizedMessage());
      initFailed = true;
    }

    if (initFailed) {
      HelpFormatter formatter = new HelpFormatter();
      formatter.printHelp(120, "java -jar labcodeset-transforms-jar-with-dependencies.jar", "", options, "", true);
    } else {
      try {
        LabcodesetResourceGenerator generator =
            new LabcodesetResourceGenerator(labcodesetFile, outputDir, loincVersion, fhirEndpoint, tokenEndpoint, clientId, clientSecret);
        generator.generateFhirResources();
      } catch (IOException e) {
        System.err.println("Failed to initialise transform " + e.getLocalizedMessage());
      }
    }
  }
}
